src/v/cluster/partition_properties_stm.cc

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/partition_properties_stm.h"
+
+#include "cluster/errc.h"
+#include "cluster/logger.h"
+#include "container/fragmented_vector.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "model/namespace.h"
+#include "model/record_batch_types.h"
+#include "raft/persisted_stm.h"
+#include "snapshot.h"
+#include "storage/log.h"
+
+#include <seastar/core/chunked_fifo.hh>
+#include <seastar/coroutine/as_future.hh>
+#include <seastar/util/defer.hh>
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <iterator>
+
+namespace cluster {
+
+partition_properties_stm::partition_properties_stm(
+  raft::consensus* raft,
+  ss::logger& logger,
+  storage::kvstore& kvstore,
+  config::binding<std::chrono::milliseconds> sync_timeout)
+  : raft::persisted_stm<raft::kvstore_backed_stm_snapshot>(
+      partition_properties_stm_snapshot, logger, raft, kvstore)
+  , _sync_timeout(std::move(sync_timeout)) {}
+
+partition_properties_stm::properties_update
+  partition_properties_stm::_default_values{
+    .writes_disabled = writes_disabled::no,
+  };
+
+ss::future<>
+partition_properties_stm::ensure_local_snapshot_exists(model::offset o) {
+    co_await raft::persisted_stm<
+      raft::kvstore_backed_stm_snapshot>::ensure_local_snapshot_exists(o);
```

Comment:
ensure_local_snapshot() calls take_local_snapshot() wondering if we can GC there? Its a bit weird to update state in what looks like a read only method.